core/src/main/java/zipkin2/reporter/internal/BoundedQueue.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright The OpenZipkin Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package zipkin2.reporter.internal;
+
+/**
+ * Multi-producer, multi-consumer queue that could be bounded by count or/and size.
+ */
+interface BoundedQueue<S> extends SpanWithSizeConsumer<S> {
+    static <S> BoundedQueue<S> create(int maxSize, int maxBytes) {
```

Comment:
@codefromthecrypt would really appreciate your early feedback here: I decided to branch off `ByteBoundedQueue` & `SizeBoundedQueue` instead of conditional logic in  `ByteBoundedQueue`, the benchmarks show slightly better throughput (plus certainly less allocations size we don't need sizes buffer anymore). 

Please let me know if that is aligned / unaligned with what you've been thinking (I have not removed any  pre-flight size checks in AsyncReporter/SpanHandler yet, didn't want to go too far).